src/odemis/acq/fastem.py

```diff
@@ -18,73 +20,398 @@
+
+        except Exception as ex:
+            # Check if any field images have already been acquired; if not => just raise the exception.
+            if len(self._fields_remaining) == len(self._roa.field_indices):
+                raise
+            # If image data was already acquired, just log a warning.
+            logging.warning("Exception during roa acquisition (after some data has already been acquired).",
+                            exc_info=True)
+            exceptions = ex  # let the caller handle the exception
+
+        finally:
+            # Remove references to the megafield once the acquisition is finished/cancelled.
+            self._fields_remaining.clear()
+
+            # Finish the megafield also if an exception was raised, in order to enable a new acquisition.
+            logging.debug("Finish megafield acquisition.")
+            dataflow.unsubscribe(self.image_received)
+
+        return self.megafield, exceptions
+
+    def acquire_roa(self, dataflow):
+        """
+        Acquire the single field images that resemble the region of acquisition (ROA, megafield image).
+        :param dataflow: (model.DataFlow) The dataflow on the detector.
+        :return: (list of DataArrays): A list of the raw image data. Each data array (entire field, thumbnail,
+                                       or zero array) represents one single field image within the ROA (megafield).
+        """
+
+        total_field_time = self._detector.frameDuration.value
+        timeout = total_field_time + 5  # TODO what margin should be used?
+
+        # Acquire all single field images, which are automatically offloaded to the external storage.
+        for field_idx in self._roa.field_indices:
+            # Reset the event that waits for the image being received (puts flag to false).
+            self._data_received.clear()
+            self.field_idx = field_idx
+            logging.debug("Acquiring field with index: %s", field_idx)
+
+            self.move_stage_to_next_tile()  # move stage to next field image position
+
+            dataflow.next(field_idx)  # acquire the next field image.
+
+            # Wait until single field image data has been received (image_received sets flag to True).
+            if not self._data_received.wait(timeout):
+                # TODO here we often timeout when actually just the offload queue is full
+                #  need to handle offload queue error differently to just wait a bit instead of timing out
+                #   -> check if finish megafield is called in finally when hitting here
+                raise TimeoutError("Timeout while waiting for field image.")
+
+            self._fields_remaining.discard(field_idx)
+
+            # Cancel the acquisition and return to the run function and handle the cancellation there.
+            # Note: The acquisition of the current single field image (tile) is still finished though.
+            if self._cancelled:
+                return
```

Comment:
Do we not handle this then in the `run()` method? So if the flag` self._cancelled` is True in the `acquire_roa()` method, we return to `run()` where it is again checked: 
```
            # In case the acquisition was cancelled, before the future returned, raise cancellation error.
            if self._cancelled:
                self._future.cancel()
                raise CancelledError()
```

Reply:
I see. Indeed, that would work then. Still, I think it's clearer to immediately raise the exception here. This way it's clear that the function is not returning normally. Especially it's strange that it wouldn't return the expected type.